src/metrics/metrics.ts

```diff
@@ -0,0 +1,214 @@
+  abortSignal?: AbortSignal,
+) {
+  const body = JSON.stringify({series: allSeries});
+  const res = await fetch(DD_DISTRIBUTION_METRIC_URL, {
+    method: 'POST',
+    headers: {
+      // Note: no compression atm.
+      'Content-Type': 'application/json',
+      'DD-API-KEY': apiKey,
+    },
+    signal: abortSignal ?? null,
+    body,
+  });
+  if (!res.ok) {
+    const maybeBody = await res.text();
+    throw new Error(
+      `unexpected response: ${res.status} ${res.statusText} body: ${maybeBody}`,
+    );
+  }
+  return res;
+}
+
+// Metrics keeps track of the set of metrics in use and flushes them
+// to a format suitable for reporting to Datadog.
+export class Metrics {
+  // We currently only support Gauge metrics.
+  private _gauges: Map<string, Gauge> = new Map();
+
+  // gauge returns a gauge with the given name. If a gauge with that name
+  // already exists, it is returned.
+  public gauge(name: string) {
+    let gauge = this._gauges.get(name);
+    if (gauge === undefined) {
+      gauge = new Gauge(name);
+      this._gauges.set(name, gauge);
+    }
+    return gauge;
+  }
+
+  // Flushes all metrics to an array of time series (plural), one DatadogSeries
+  // per metric.
+  public flush(): DatadogSeries[] {
+    const allSeries: DatadogSeries[] = [];
+    for (const gauge of this._gauges.values()) {
+      const series = gauge.flush();
+      allSeries.push(series);
+    }
+    return allSeries;
+  }
+}
+
+// These two types are Datadog's. Yeah, I don't like them either.
+// DatadogSeries is a time series of points for a single metric.
+export type DatadogSeries = {
+  metric: string; // We call this 'name' bc 'metric' is overloaded in code.
+  points: DatadogPoint[];
+};
+// A point is a second-resolution timestamp and a set of values for that
+// timestamp. A point represents exactly one second in time and the values
+// are those recorded for that second. The first element of this array
+// is the timestamp and the second element is an array of values.
+export type DatadogPoint = [number, number[]];
+
+function newDatadagPoint(ts: number, value: number): DatadogPoint {
+  return [ts, [value]];
+}
+
+// Gauge is a metric type that represents a single value that can go up and
+// down. It's typically used to track discrete values or counts eg the number
+// of active users, number of connections, cpu load, etc. A gauge retains
+// its value when flushed.
+//
+// We use a Gauge to sample at the client. If we are interested in tracking
+// a metric value *per client*, the client can note the latest value in
+// a Gauge metric. The metric is periodically reported to Datadog. On the
+// server, we graph the value of the metric rolled up over the periodic
+// reporting period, that is, counted over a span of time equal to the
+// reporting period. The result is ~one point per client per reporting
+// period.
+export class Gauge {
+  private _name: string;
+  private _point: DatadogPoint | undefined;
```

Comment:
please assign a value. TS should have warned about this one but 

I could not find a TS bug about this so I filed a new one

https://github.com/microsoft/TypeScript/issues/52259

Reply:
Think I fixed it?